### PR TITLE
Use ES aliases instead of index names in queries

### PR DESCRIPTION
**Pull request by sravfeyn**
This is only open to get any feedback and not ready for merge.

We are currently querying ES via index names. Querying via alias is supported and will be needed for ICDS where we are thinking to have multiple indices under one alias. This PR doesn't implement any multiple index support, but as a starting point simply queries ES using alias instead of index in all places that I could find. I have added a logger to find any places that I missed where an index is directly queried instead of an alias.

My plan for this PR is
- to get tests passing
- test it on staging with `ENABLE_ES_INTERFACE_LOGGING`
- if those two go okay, I will open this PR for merge.  

After this my plan to actually add multiple index support for single alias. My tentative plan is
 
- add an additional `write_alias` attribute to `ElasticsearchIndexInfo` and use that for all write operations.
- add a validation to check that `write_alias` points to only single index
- rename current `alias` attribute of `ElasticsearchIndexInfo` to read_alias and use that for all `read` operations.

This will enable individual envs to manage their ES indices outside of HQ with custom settings, rolling indices etc

### CODE REVIEW

File: corehq/elastic.py
Comment (snopoke) [inline corehq/elastic.py]: :heart:

File: corehq/util/es/interface.py
Comment (snopoke) [inline corehq/util/es/interface.py]:
```suggestion
                logger.info("Found a use case where index is queries instead of alias: %s", index_or_alias)
```

File: corehq/util/es/interface.py
Comment (snopoke) [reply corehq/util/es/interface.py]:
```suggestion
                logger.info(traceback.format_stack())
```

File: corehq/util/es/interface.py
Comment (snopoke) [inline corehq/util/es/interface.py]: rename `index` to `index_alias` to be consistent with other methods

**Review by snopoke — COMMENTED** [review]
Looks good @sravfeyn

File: corehq/util/es/interface.py
Comment (sravfeyn) [reply corehq/util/es/interface.py]: (Done)

**Comment by sravfeyn** [comment]
@snopoke  @dannyroberts 

FYI, I hit a small blocker on this. It impacts the deploy/reindex workflow, so I am flagging you both

The issue is that this PR makes the reindexer code also use aliases instead of index names. Since our workflow when editing an Elasticsearch mapping is to only create a new index in a private release, but not assign any alias, the reindexer code writes to old index instead of new because it continues to use the old alias pointing to old  index. There are two options here to resolve this. 

One is to update our workflow of changing ES mapping to also assign a separate ES alias, let's say, xforms_new pointing to new index, this way the reindexer will use the new alias pointing to new index. Once index is ready, flip the alias xforms from old index to new index (which should happen automatically in the code deploy which has flip_all command).

Second option is to make all the reindexer code to explicitly use index name instead of alias. The issue with this is that it needs a flag to be passed in the call chain of Reindexer -> Pillow_Or_processor -> ElasticsearchInterface. Luckily, this needs change in [just one place](https://github.com/dimagi/commcare-hq/pull/26733/commits/1844cdb9d86b3c3c5435a167cb0543077fb4cf00#diff-37cbd762f6bb728a469145592aff76b0R211) for reindexers that use ResumableBulkElasticPillowReindexer. So I converted some reindexers that don't use that reindexer to use it in this commit https://github.com/dimagi/commcare-hq/pull/26733/commits/1844cdb9d86b3c3c5435a167cb0543077fb4cf00 but there are two more Reindexers which need a lot more effort to refactor to use ResumableBulkElasticPillowReindexer .

So, I am leaning to go with option 1 since that avoids using index-name at all and is much minimal change than the 2nd option. What do you guys think?

**Comment by snopoke** [comment]
> but there are two more Reindexers which need a lot more effort to refactor to use ResumableBulkElasticPillowReindexer.

which ones are these?

**Comment by sravfeyn** [comment]
> which ones are these?
The two are ReportCaseReindexerFactory and the ReportFormReindexerFactory. I have tried this in one of the latest commits, but reverted as it's quite more complicated than that.

> My initial feeling is that we should change the reindexers to use the index names directly however that could be wrong if the aliases get set up with lifecycle / rollover policies in which case you'd want to write to the alias directly. So I think you may be right that creating a new alias when reindexing is the way to go.

Yeah, I think this feels the right way, I have updated the reindexer code to set alias also and use it. I have updated the relevant ES reindexing docs to reflect this.

**Comment by sravfeyn** [comment]
@dannyroberts This will affect all envs, can you confirm if this is okay for other envs as well? Docs https://github.com/dimagi/commcare-hq/pull/26733/files#diff-55ae021bb3841b645117710c304ad227

File: docs/elasticsearch.rst
Comment (dannyroberts) [inline docs/elasticsearch.rst:70]: nit
```suggestion
that new mapping. Updating the index name in the mapping file triggers HQ to
create the new index with the new mapping and reindex all data, so you'll
have to update the index hash and alias at the top of the mapping file.
```

File: docs/elasticsearch.rst
Comment (dannyroberts) [inline docs/elasticsearch.rst:76]: I'm kind of just catching up here. Can you explain why we need to change the alias as well as the index name? I thought the whole point of using an alias and an index name was that the index name could change while keeping the alias the same

Comment (dannyroberts) [reply docs/elasticsearch.rst:76]: OK sorry, I see the large comment above about this.

Looks like this also felt wrong to you guys but you couldn't figure out a way to use the index that still had the right behavior?

Comment (sravfeyn) [reply docs/elasticsearch.rst:76]: Yeah, this doesn't feel completely right but is better than other options. This is discussed in [comments following this](https://github.com/dimagi/commcare-hq/pull/26733#issuecomment-610909316)

All of the current code uses index names directly rather than aliases, this PR starts making use of the aliases. But our index mapping change workflow includes building a new index with HQ itself, so without creating a separate alias during the index phase, there isn't any way to prevent writing to the old index itself.

Comment (snopoke) [reply docs/elasticsearch.rst:76]: @dannyroberts I don't think it's that we couldn't figure out a way to use the index. An alias is designed to allow you to run searches across multiple indexes which is what we'd like to start using it for (e.g. monthly form indexes). We have been using it in CommCare as a way to 'flip' from one index to another after reindex (though we haven't actually used the aliases in code for a long time).

**Comment by sravfeyn** [comment]
I have tested this on staging by loading few ES dependent features and making sure that only aliases are queried and not indices (by looking at HTTP logs on ES obtained by tcpdumps and making sure that this [soft-assert log](https://github.com/dimagi/commcare-hq/pull/26733/files#diff-35449daeec5614e63f3a53d79459660aR14) isn't triggered). So this is ready for merge pending any further code review. I am going to convert that soft assert log into a hard failure in a separate PR in future after a week of this being live on other envs without any issues.

All of HQ with this PR uses aliases for reading/writing, and uses direct index names for index management and pillow checkpointing.

Pinging for a final code review on this @dannyroberts @esoergel @proteusvacuum @orangejenny  (cc @snopoke)

File: corehq/apps/es/case_search.py
Comment (sravfeyn) [reply corehq/apps/es/case_search.py:34]: It could be. This should be one of the keys in `ES_META`, it's neither an alias, nor and index name. Since `CASE_SEARCH_ALIAS` may no longer evaluate to this value, I had to change it. It's consistent with all other `HQESQuery` classes, so I let this be.

File: corehq/apps/domain/tests/test_domain_calculated_properties.py
Comment (sravfeyn) [reply corehq/apps/domain/tests/test_domain_calculated_properties.py:49]: It shouldn't be an issue, since the test is not about ES, but to be consistent, I will update it

File: corehq/ex-submodules/pillowtop/es_utils.py
Comment (sravfeyn) [reply corehq/ex-submodules/pillowtop/es_utils.py:143]: Yeah, missed that. Actually the parent function is an orphan, so I will remove it altogether.

**Comment by sravfeyn** [comment]
Hey guys  @snopoke @esoergel @dannyroberts , can I get a review on this (I think you have already reviewed most of it). This was blocked due to tests taking lot of time due to the auto-create-index I added previously. I have now fixed that issue in [this other PR](https://github.com/dimagi/commcare-hq/pull/27614) so this is now good to go pending a review.

File: corehq/apps/receiverwrapper/tests/__init__.py
Comment (dannyroberts) [inline corehq/apps/receiverwrapper/tests/__init__.py:9]: nit: doesn't seem like this needs to be in the changeset

File: docs/elasticsearch.rst
Comment (dannyroberts) [reply docs/elasticsearch.rst:28]: nit: bump

File: docs/elasticsearch.rst
Comment (dannyroberts) [reply docs/elasticsearch.rst:70]: nit: bump

**Review by dannyroberts — APPROVED** [review]
Had a few small comments, but overall it looks good

**Comment by dannyroberts** [comment]
Looks like there are still issues with the tests on this PR and that's what the other PR's supposed to solve? Seems worth merging that other PR first and the merging master into this and making sure tests pass after that?

**Comment by sravfeyn** [comment]
Thanks Danny. I made another PR that should fix that issue https://github.com/dimagi/commcare-hq/pull/27849/.

**Comment by sravfeyn** [comment]
Build passed. Testing one last time on staging before merging.

**Comment by sravfeyn** [comment]
@snopoke @dannyroberts I have added few commits to fix an issue in rolling out an index change that require a review.

**Comment by sravfeyn** [comment]
Bumping for review on recently added commits.

**Comment by sravfeyn** [comment]
@snopoke @dannyroberts Bumping again for review on last 3 commits

File: corehq/ex-submodules/pillowtop/tests/utils.py
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/tests/utils.py:37]: does this work for both versions?

File: docs/elasticsearch.rst
Comment (snopoke) [inline docs/elasticsearch.rst:77]: when the alias is renamed in code will that automatically get updated in ES? Will the old (`xforms_new`) alias still exist or does it get removed?

**Comment by sravfeyn** [comment]
@dannyroberts @snopoke Is this good to go?

**Comment by snopoke** [comment]
@sravfeyn still some outstanding comments

**Comment by sravfeyn** [comment]
@snopoke Sorry, which ones Simon?

**Comment by snopoke** [comment]
@sravfeyn https://github.com/dimagi/commcare-hq/pull/26733#discussion_r448228772 and https://github.com/dimagi/commcare-hq/pull/26733#discussion_r448221878

**Comment by sravfeyn** [comment]
I am not sure why you are not able to see my responses made couple days back.

Screenshot 2020-07-07 at 4 58 11 PM

**Comment by snopoke** [comment]
@sravfeyn thanks for the responses, your comments are still pending, you need to click 'submit review' (I've been bitten by that many times).

File: docs/elasticsearch.rst
Comment (snopoke) [reply docs/elasticsearch.rst:77]: If the old alias remains then I would suggest a different suffix other than `_new` like the date or something

Comment (sravfeyn) [reply docs/elasticsearch.rst:77]: Good idea, I have updated the docs.

**Comment by sravfeyn** [comment]
> you need to click 'submit review' (I've been bitten by that many times).
Ah, thanks!

**Comment by sravfeyn** [comment]
@snopoke I have responded to the comments (hope you are able to see them) and update the docs per your comment.

File: docs/elasticsearch.rst
Comment (snopoke) [reply docs/elasticsearch.rst:77]: I don't see the changes

**Review by snopoke — APPROVED** [review]

**Comment by snopoke** [comment]
@sravfeyn have these changes been tested on staging at all?

**Comment by sravfeyn** [comment]
This is in QA now

**Comment by sravfeyn** [comment]
QA is passed on this https://dimagi-dev.atlassian.net/browse/QA-1532

@snopoke think we can merge this now?

**Review by snopoke — APPROVED** [review]

**Comment by sravfeyn** [comment]
Hello @dimagi/team-commcare-hq 

This is a pretty big change in Elasticsearch, so wanted to make an announcement. From this PR on, HQ code will use Elasticsearch aliases to interact with Elasticsearch instead of index names as it previously did. There is also a change in the workflow of changing ES index mappings, please read it here https://github.com/dimagi/commcare-hq/pull/26733/files#diff-55ae021bb3841b645117710c304ad227

This is QA and review pass. I am going to merge it tomorrow (to give some time for people to be aware and raise any concerns if any)

**Comment by sravfeyn** [comment]
This is well tested but just to be safe, if any issues with pillows do come up form this PR, this can be reverted. We just need to rewind pillows https://github.com/dimagi/commcare-hq/pull/13154/files#diff-5e1c4d805a2ca1169c83e9e2446c318eR76

**Pull request merged**
